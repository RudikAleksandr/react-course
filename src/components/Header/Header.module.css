
.header {
  background: linear-gradient(to top,rgba(0,0,0,.7) 0,rgba(0,0,0,.7) 100%), url('/src/assets/images/netflixMovies.jpg');
  min-height: 400px;
  width: 100%;
}

.netflixLink {
  color: #f65261;
  font-size: 1.5rem;
}

.netflixLink:hover {
  color: #f65261;
}

.header .addMovieButton {
  color: #f65261;
  font-size: 1.2rem;
  font-weight: 500;
  text-transform: uppercase;
  background-color:rgba(85, 85, 85, 0.8);
  float: right;
  border: none;
}

.header .addMovieButton:hover  {
  background-color:#555;
  color: #f65261;
}
