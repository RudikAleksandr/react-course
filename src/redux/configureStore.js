import { configureStore as configureStoreToolkit } from '@reduxjs/toolkit';
import thunkMiddleware from 'redux-thunk';
import moviesReducer from './moviesSlice';

const configureStore = (preloadedState) => (
  configureStoreToolkit({
    preloadedState,
    reducer: {
      movies: moviesReducer,
    },
    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
  })
);

export default configureStore;
