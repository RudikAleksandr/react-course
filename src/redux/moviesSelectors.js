
const selectListMovies = (state) => state.movies.moviesOrder.map(id => state.movies.movies[id]);

const selectSearch = (state) => state.movies.search;

const selectActiveMovieGenre = (state) => state.movies.activeMovieGenre;

const selectActiveSortType = (state) => state.movies.activeSortType;

export {
  selectListMovies,
  selectSearch,
  selectActiveMovieGenre,
  selectActiveSortType
};