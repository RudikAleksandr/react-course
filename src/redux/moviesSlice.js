import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import * as moviesAPI from '../api/moviesAPI';
import { groupByKey } from '../utils/grouping';

export const initialState = {
  movies: {},
  moviesOrder: [],
  search: '',
  activeMovieGenre: '',
  activeSortType: 'release_date'
};

export const getListMovies = createAsyncThunk(
  'movies/getListMovies',
  moviesAPI.getListMovies
);

export const getMovie = createAsyncThunk(
  'movies/getMovie',
  moviesAPI.getMovie
);

export const updateMovie = createAsyncThunk(
  'movies/updateMovie',
  moviesAPI.updateMovie
);

export const createMovie = createAsyncThunk(
  'movies/createMovie',
  moviesAPI.createMovie
);

export const deleteMovie = createAsyncThunk(
  'movies/deleteMovie',
  moviesAPI.deleteMovie
);

const moviesSlice = createSlice({
  name: 'movies',
  initialState,
  reducers: {
    setSearch: (state, { payload }) => {
      state.search = payload;
    },
    setActiveMovieGenre: (state, { payload }) => {
      state.activeMovieGenre = payload;
    },
    setActiveSortType: (state, { payload }) => {
      state.activeSortType = payload;
    }
  },
  extraReducers: {
    [getListMovies.fulfilled]: (state, { payload }) => {
      state.movies = groupByKey('id', payload.data);
      state.moviesOrder = payload.data.map((movie) => movie.id);
    },
    [updateMovie.fulfilled]: (state, { payload }) => {
      state.movies[payload.id] = payload;
    },
  },
});

export const {
  setSearch,
  setActiveMovieGenre,
  setActiveSortType
} = moviesSlice.actions;

export default moviesSlice.reducer;