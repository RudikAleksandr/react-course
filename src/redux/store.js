import { configureStore } from '@reduxjs/toolkit';
import moviesReducer from './moviesSlice';
import thunkMiddleware from 'redux-thunk'

const store = configureStore({
  reducer: {
    movies: moviesReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
});

export default store;