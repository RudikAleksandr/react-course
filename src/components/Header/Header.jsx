import React, { useCallback, useEffect, useReducer } from 'react';
import { Row, Col, Button } from 'react-bootstrap';
import { Link, useHistory } from "react-router-dom";
import { useDispatch } from 'react-redux';
import { useLocation } from "react-router-dom";
import { setSearch } from '../../redux/moviesSlice';
import MovieFormModal from '../MovieFormModal';
import SearchMovies from '../SearchMovies';
import cn from './Header.module.css';

// Custom Hook
export const useToggle = (initialValue = false) => {
  return useReducer((state) => !state, initialValue);
}

const Header = () => {
  const history = useHistory();
  const searchQuery = new URLSearchParams(useLocation().search).get('query');

  const [isOpenCreateModal, toggleIsOpen] = useToggle();
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(setSearch(searchQuery || ''));
  }, [searchQuery, dispatch]);

  const handleSearchMovies = useCallback((search) => {
    history.push(`/search?query=${search}`);
  }, [history])

  const handleToggleCreateModal = useCallback(() => {
    toggleIsOpen();
  }, [toggleIsOpen]);

  return (
    <Col className={cn.header}>
      <Row className="justify-content-between align-items-end mb-5 ml-5 mr-5 mt-3">
        <Link className={cn.netflixLink} to="/"><b>netflix</b>roulette</Link>
        <Button
          variant=""
          className={cn.addMovieButton}
          onClick={handleToggleCreateModal}
        >
          + Add movie
        </Button>
      </Row>
      <SearchMovies onSearch={handleSearchMovies} />
      {isOpenCreateModal &&
        <MovieFormModal onHide={handleToggleCreateModal} />
      }
    </Col>
  );
}

export default Header;
