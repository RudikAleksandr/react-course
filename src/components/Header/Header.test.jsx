import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter, Route } from 'react-router';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import userEvent from '@testing-library/user-event';
import Header from './Header';
import { initialState } from '../../redux/moviesSlice';

const getComponent = (store) => (
  <MemoryRouter>
    <Route>
      <Provider store={store}>
        <Header />
      </Provider>
    </Route>
  </MemoryRouter>
);

describe('Header component', () => {
  test('render Header snapshot', () => {
    const store = configureStore()({ movies: initialState });
    const component = getComponent(store);
    const { asFragment } = render(component);

    expect(asFragment(component)).toMatchSnapshot();
  });

  test('should open create movie modal', async () => {
    const store = configureStore()({ movies: initialState });
    const component = getComponent(store);

    const { findByTestId, getByText, queryByTestId } = render(component);

    expect(queryByTestId('movieFormModal')).not.toBeInTheDocument();

    userEvent.click(getByText('+ Add movie'), null);

    expect(await findByTestId('movieFormModal')).toBeInTheDocument();
  });
});
