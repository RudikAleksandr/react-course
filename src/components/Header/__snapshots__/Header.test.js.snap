// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component render Header snapshot 1`] = `
<DocumentFragment>
  <div
    class="header col"
  >
    <div
      class="justify-content-between align-items-end mb-5 ml-5 mr-5 mt-3 row"
    >
      <a
        class="netflixLink"
        href="/"
      >
        <b>
          netflix
        </b>
        roulette
      </a>
      <button
        class="addMovieButton btn"
        type="button"
      >
        + Add movie
      </button>
    </div>
    <div
      class="justify-content-center searchMovies row"
    >
      <div
        class="col-md-7 col-11"
      >
        <h1
          class="findMovieTitle"
        >
          Find your movie
        </h1>
      </div>
      <div
        class="mt-4 col-md-7 col-11"
      >
        <div
          class="justify-content-center row"
        >
          <div
            class="col-xl-8 col-8"
            xxl="9"
          >
            <div
              class="form-group"
            >
              <input
                class="searchInput form-control"
                placeholder="What do you want to watch?"
                value=""
              />
            </div>
          </div>
          <div
            class="pl-0 col-4"
          >
            <button
              class="btn btn-primary"
              type="button"
            >
              Search
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
